The mostly agreed parts of the evidence-verification-min PR (#91)
ietf-rats-wg/draft-ietf-rats-corim: draft-ietf-rats-corim.md

  The `$$concise-bom-tag-extension` extension socket is empty in this
  specification.

# CoRIM-based Evidence Verification

The verification procedure is divided into three separate phases:

* Appraisal Context initialisation
* Evidence collection
* Evidence appraisal

At a few well-defined points in the procedure, the Verifier behaviour will
depend on the specific CoRIM profile.
Each CoRIM profile MUST provide a description of the expected Verifier behavior
for each of those well-defined points.

> Note that what follows describes a simplified implementation.

Reviewer: ```suggestion
Note that what follows describes a simplified algorithm for verification.
```

Reviewer 2: Note that what follows describes a simplified algorithm.

Reviewer: @nedmsmith : Me and Andrew thinks, "standard algorithm is neat and is specified in the standard document, so works for us atleast!

Reviewer 2: WFM

Author: standard algorithm it is

Verifiers claiming compliance with this specification MUST exhibit the same
externally visible behavior as described here,
they are not required to use the same internal data structures.
For example, it is expected that the resources used during the initialisation
phase can be amortised across multiple appraisals.

## Appraisal Context initialisation

The goal of the initialisation phase is to load the CoRIM Appraisal Context
with objects such as tags (CoMID, CoSWID, etc.) from CoRIM files,
cryptographic validation key material (e.g., raw public keys, root certificates,
intermediate CA certificate chains), etc. that will be used in the subsequent
Evidence Appraisal phase.

### CoRIM Selection

All available CoRIMs are collected.

> A Verifier may be pre-configured with a large number of tags describing many

Reviewer: ```suggestion
A Verifier may be pre-configured with a large number of CoRIMs describing many
```

types of device.
All CoRIMs are loaded at this stage, later stages will select the CoRIMs
appropriate to the Evidence Appraisal step.

CoRIMs that are not within their validity period, or that cannot be associated
with an authenticated and authorised source MUST be discarded.

CoRIM that are secured by a cryptographic mechanism such as a signature which
does not pass validation MUST be discarded.

Other selection criteria MAY be applied.

> For example, if the Evidence format is known in advance, tags that do not match

Reviewer: ```suggestion
For example, if the Evidence format is known in advance, CoRIMs that do not match
```

> For example, if the Evidence format is known in advance, tags that do not match
> the expected profile can be readily discarded.

Reviewer: ```suggestion
For example, if the Evidence format is known in advance, CoRIMs using a profile that is not understood by a Verifier can be readily discarded.
```

Reviewer 2: Given matching semantics are encoded by triples, shouldn't the language here focus on triples?

Author: Using Yogesh's proposal

> The selection process MUST yield at least one usable tag.

Reviewer 2: At this point, the RIM is discarded (except that the RIM's signer is remembered - possibly by injecting `authorized-by` attributes). The next level processing applies to tags. The current text seems correct.

> The selection process MUST yield at least one usable tag.

Author: The verification process could activate tags which are not CoMIDs, so I think referring to tags is appropriate here.

### CoBOM Extraction

> All the available Concise Bill Of Material (CoBOMs) tags are then collected

Reviewer: Can you please help me clarify:

For a given profile, only one active CoBOM is expected at the point of verification.

So does this section assumes that the set of CoBOMs pertain to all the CoBOMs (for a variety of attester types) and thus all profiles which Verifier handles?
If such is the assumption, then one must explain how to filter the `active` CoBOM relevant to a specific profile.

Reviewer 2: CoBOMs can contain tags that are not in the RIM that immediately contains the CoBOM. Dependent RIMS (containing these tags) may, themselves, have CoBOMs that reference tags in the immediate RIM (or yet other RIMS), and so forth. 
Presumably, there is a "top level" CoBOM that correctly describes the "device" that is being attested. The question for Verifiers is how does it know it has the complete description (via tags) of the device? Which CoBOM is the 'root' CoBOM?

from the selected CoRIMs.

> 

Author: I added a paragraph in the section above telling the verifier to discard invalid CoBOMs before using them.

The Verifier MUST activate all tags referenced by a CoBOM.

After the Verifier has processed all CoBOMs it MUST discard any tags which have
not been activated by a CoBOM.

> ### Tags Identification and Validation

Author: I think the two steps are separate, so it is clearer to describe them sequentially. If you feel strongly then we can reconsider.

The Verifier chooses tags -- including Concise Module ID Tags (CoMID, {{sec-comid}}),
Concise Software ID Tags (CoSWID, {{-coswid}}),

> and/or Concise Trust Anchor Stores (CoTS, {{?I-D.ietf-rats-concise-ta-stores}}) --

Reviewer 3: use the already defined alias (`ta-store`) or it will break:

```suggestion
and/or Concise Trust Anchor Stores (CoTS, {{-ta-store}}) --
```

> from the selected CoRIMs.

Reviewer: ```suggestion
from the selected CoRIMs activated by a CoBOM.
```

Reviewer 3: @yogeshbdeshpande: the text you suggest seems redundant given we are past the CoBOM processing step.  (But see also my proposal to merge the subsections.)

Author: I agree, no need to state that twice.

The Verifier MUST discard all tags which are not syntactically and semantically
valid.
In particular, any cross-referenced triples (e.g., CoMID-CoSWID linking triples)
MUST be successfully resolved.

### Appraisal Context Construction

All of the validated and potentially useful tags are loaded into the Appraisal Context.
Each tag is loaded into the Appraisal Context as a single unit, later stages

> will accept or ignore the whole tag.

Reviewer 3: unclear what this means.  Is there a typo?

Reviewer 2: possibly: "...will accept or ignore the tag in its entirety."

Author: We need to make a decision about whether tag boundaries are used as part of matching. Recent conversations in the group were going in the direction where they are ignored, but I think it may simplifiy things if we keep them. I will turn this into an issue.

This concludes the initialisation phase.

## Evidence Collection

In the evidence collection phase the Verifier communicates with attesters to
collect evidence.

The first part of the Evidence collection phase does not perform any
cryptographic validation.
This allows Verifiers to use untrusted code for their initial Evidence collection.

The results of the evidence collection are protocol specific data and transcripts
which are used as input to appraisal by the Verifier.

### Cryptographic validation of Evidence

If the authenticity of Evidence is secured by a cryptographic mechanism such as
a signature, the first step in the Evidence Appraisal is to perform
cryptographic validation of the Evidence.

The exact cryptographic signature validation mechanics depend on the specific
Evidence collection protocol.

For example:
In DICE, a proof of liveness is performed on the final key in the certificate
chain.
If this passes then a suitable certification path anchored on a trusted root
certificate is looked up -- e.g., based on linking information obtained from

> the DeviceID certificate (see Section 9.2.1 of {{DICE-Layering-Architecture}}) --

Reviewer 3: remember to add the bib entry for DICE-Layering-Architecture or it won't compile.

Reviewer 2: DICE Layering Architecture, Version 1.0, Revision 0.19, July 23, 2020. https://trustedcomputinggroup.org/wp-content/uploads/DICE-Layering-Architecture-r19_pub.pdf

in the Appraisal Context.  If found, then usual X.509 certificate validation
is performed.
In PSA, the verification public key is looked up in the appraisal context using

> the `euid` claim found in the PSA claims-set (see {{Section 4.2.1 of -psa-token}}).

Reviewer 3: ```suggestion
the `ueid` claim found in the PSA claims-set (see {{Section 4.2.1 of -psa-token}}).
```

If found, COSE Sign1 verification is performed accordingly.

Independent of the specific integrity protection method used, the integrity of
Evidence MUST be successfully verified.

> A CoRIM profile MUST describe:
>
> * How cryptographic verification key material is represented (e.g., using Attestation Keys triples, or CoTS tags)
> * How key material is associated with the Attesting Environment
> * How the Attesting Environment is identified in Evidence

### The Accepted Claims Set

At the end of the Evidence collection process evidence has been converted into

> a format suitable for appraisal. This document describes an `accepted-claims-set`

Reviewer: ```suggestion
a format suitable for appraisal, known as `accepted-claims-set`. This document describes an `accepted-claims-set`
```

Reviewer 3: I don't think the specific suggestion works (esp. in the flow with the rest of the sentence).  But I agree we need to be a bit more precise here about what the `accepted-claims-set` is supposed to represent in our "exemplary" verification flow.  Maybe:
```suggestion
a format suitable for appraisal. To this end, this document describes an `accepted-claims-set`
```

format and the algorithms used to compare it against CoMID reference values.

    accepted-claims-set = {​
        &(ce.evidence-triples: 0) => [ + reference-triple-record ]​
      ? &(ce.identity-triples: 1) => [ + identity-triple-record ]​
      ? &(ce.dependency-triples: 2) => [ + domain-dependency-triple-record ]
      ? &(ce.domain-membership-triples: 3) => [ + domain-membership-triple-record ]
      ? &(ce.coswid-triples: 4) => [ + ev-coswid-triple-record ]
      * $$ev-triples-map-extension​

>     }

Reviewer 3: we probably want to move this together with the rest of the CDDL (to apply automatic validation & consistency checks).  If you'd rather make it in a separate PR, please raise an issue from this comment.

Author: I agree, this should go into a separate file (and be formatted correctly). Can I do that with a separate change after we accept this one?

Verifiers are not required to use this as their internal state, but for the
purposes of this document a sample Verifier is discussed which uses this format.

The Accepted Claims Set will be matched against CoMID reference values, as per
the appraisal policy of the Verifier.
This document describes an example evidence structure which can be easily
matched against these reference values.
Each set of evidence contains an `environment-map` providing a namespace, and
a non empty `measurement-values-map`.

Each entry in the `measurement-values-map` is a separate piece of evidence

> describing the environment named by the `environment-map`.

Reviewer: ```suggestion
describing the measurements associated to the environment identified in the `environment-map`.
```

 An Attester can provide multiple `environment-map`s each containing a

>  `measurement-values-map` with one entry;  a single `environment-map` containing

Reviewer: 100 % agree with statement on line 1360 and hence we should support this expression natively in the base CDDL.

 multiple entries in its `measurement-values-map`; or a combination of
 these approaches.

If evidence from different sources has the same `environment-map` then the
`measurement-values-map`s are merged.

> If both measurement-value-maps being merged contain the same key then the

Reviewer: ```suggestion
If measurement-value-maps being merged contain the same key then the
```

> values associated with that key MUST be binary identical.

Reviewer: Also can you please clarify what you meant by `key` here? is this referring to a key-value pair concept or something else?

### Accepted Claims Set Initialisation

The Accepted Claims Set is initialised with cryptographically verified Evidence
from the Attestation Environments.

> A CoRIM profile MUST describe:
>
> * How evidence is converted to a format suitable for appraisal

> Section {sec-dice-spdm} provides information on how evidence collected using

Reviewer 3: refs are `{{...}}`.  Besides, the processor is smart enough to prepend Section, Figure depending on the anchor.  So what you need is:
```suggestion
{{sec-dice-spdm}} provides information on how evidence collected using
```

match the Attester. Tags which match are accepted, and the Accepted Claims Map
is extended using Endorsements etc. from the accepted tags.

[^issue]: Content missing. Tracked at https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/71

### Comparing and processing CoMID tags

[^issue]: Content missing. Tracked at https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/71

### Matching Evidence against Reference Values

[^issue]: Content missing. Tracked at https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/71

### Adding CoMID Endorsed Values to the Accepted Claims Set

[^issue]: Content missing. Tracked at https://github.com/ietf-rats-wg/draft-ietf-rats-corim/issues/71

## Adding DICE/SPDM evidence to the Accepted Claims Set {#sec-dice-spdm}

This section defines how evidence from DICE and/or SPDM is transformed into a
format where it can be added to an accepted claims set.
A Verifier supporting DICE/SPDM format evidence should implement this section.

### Transforming SPDM Evidence to a format usable for matching

[Evidence Binding For SPDM](TCG_SPDM-TBD) describes the process by which

> evidence in a SPDM MEASUREMENTS response is converted to Evidence suitable for

Author: This section is talking about DICE. The SPDM section is later (unless your comment was moved)
